### PR TITLE
dm: turn on/off relay by source config

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 

we re-support old behavour. if user set `enable-relay: true` in source config, we'll use relay replication for the bound worker.

note that `start-relay -s source workername` (whether executed now or upgraded from metadata of old version) on a `enable-relay: true` source is undefined behaviour.

### What is changed and how it works?
⬆️

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported variable/fields change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### Release note 

```release-note
`enable-relay: true` in source config takes effect again
```

### CODE REVIEW

File: dm/dm/worker/server.go [hunk 1]
```diff
@@ -340,6 +340,10 @@ func (s *Server) observeRelayConfig(ctx context.Context, rev int64) error {
 					}
 					rev = rev1
 					if relaySource == nil {
+						if w := s.getWorker(false); w != nil && w.startedRelayBySourceCfg {
```
Comment (lichunzhu) [inline dm/dm/worker/server.go:343]: `s` is not locked here. Should we use lock?
```suggestion
						if w := s.getWorker(true); w != nil && w.startedRelayBySourceCfg {
```

File: dm/dm/worker/server.go [hunk 2]
```diff
@@ -340,6 +340,10 @@ func (s *Server) observeRelayConfig(ctx context.Context, rev int64) error {
...
+							break
+						}
+						log.L().Info("didn't found relay config after etcd retryable error")
```
Comment (lichunzhu) [inline dm/dm/worker/server.go:346]: Just suggestion
```suggestion
						log.L().Info("didn't found relay config after etcd retryable error. Will start to stop relay now")
```

**Review by lichunzhu — APPROVED** [review]
Do we have an integration test for this? If not, we can add that in a later PR

**Comment by lance6716** [comment]
> Do we have an integration test for this? If not, we can add that in a later PR

Our old CI has set many `enable-relay: true`. I changed many "Running" number check in test scripts because now "Running" relay unit is started, so I think the current integration test should be enough.

File: dm/dm/worker/server.go (hunk 2 above)
Comment (lance6716) [reply dm/dm/worker/server.go]: above comments should be fixed in https://github.com/pingcap/ticdc/pull/3190/commits/4d349c7e3dd1ec5592b55849e75a35a9ae8f7c77

File: dm/dm/worker/server.go [hunk 3]
```diff
@@ -340,6 +340,10 @@ func (s *Server) observeRelayConfig(ctx context.Context, rev int64) error {
 					}
 					rev = rev1
 					if relaySource == nil {
+						if w := s.getWorker(true); w != nil && w.startedRelayBySourceCfg {
```
Comment (Ehco1996) [inline dm/dm/worker/server.go:343]: maybe some retry logic can move in to `w.handleRelayStage`, for retry, jus call handleRelayStage again?

Comment (Ehco1996) [reply dm/dm/worker/server.go:343]: don't have to change it in this pr, I'm just making a slight suggestion

Comment (lance6716) [reply dm/dm/worker/server.go:343]: what's the relation of this function and handleRelayStage?

Comment (Ehco1996) [reply dm/dm/worker/server.go:343]: sorry, it's `handleRelayConfig` not `handleRelayStage`
i think from L324 to L377 is retry logic for `handleRelayConfig`, so maybe it's beterr handle retry in `handleRelayConfig`

**Review by Ehco1996 — APPROVED** [review]
LGTM 🉑 please fix the confilcts

**Comment by GMHDBJD** [comment]
Need update upstream-switch case https://github.com/pingcap/ticdc/runs/4209093582?check_suite_focus=true

**Comment by GMHDBJD** [comment]
need update https://github.com/pingcap/ticdc/runs/4209093582?check_suite_focus=true @lance6716

**Pull request merged**